CI: Remove Bazel usage from the pipeline and port sample-values test to bash build system

viccuad opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

- Substitute Bazel with Bash build system for lint & build jobs in the pipeline.
- Use `splatform/catapult` container image instead of `thulioassis/bazel-docker-image` in those jobs.
- Port functionality of `bazel test //deploy/helm/kubecf:values_doc_test` into `make kubecf-build`.
- Only print errors when testing the sample `values.yaml`, to leave `make kubecf-build` output unchanged.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Part of https://github.com/cloudfoundry-incubator/kubecf/issues/1010.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- An important detail to include is the version of the used cf-operator -->

Testing in https://concourse.suse.dev/teams/main/pipelines/kubecf-rmbazel
lint, build, publish should be green in master (maybe publish will fail because the tar is already there). The pr tab is used for updating this PR's github statuses.

Manual tests for the change in testing sample values doc tests have been performed:
- Bazel:
  * true positive by running bazel test //deploy/helm/kubecf:values_doc_test
  * true negative by running bazel test //deploy/helm/kubecf:values_doc_test and
    removing comment for system_domain in deploy/helm/kubecf/values.yaml
- Bash build scripts:
  * true positive by running `make kubecf-build`
  * true negative by running `make kubecf-build` and removing comment for
    system_domain in deploy/helm/kubecf/values.yaml


## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code has security implications.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

viccuad commented:
Testing blocked by  #1095, hence, opened as draft.

viccuad commented:
Can be tested by approving, and then pinning this PR in https://concourse.suse.dev/teams/main/pipelines/kubecf-rmbazel?group=pr. This will update the github statuses here.

viccuad commented:
The bash build system downloads on its own shellcheck, yamllint, bosh, instead of using any of those binaries from the CI catapult image: https://concourse.suse.dev/teams/main/pipelines/kubecf-rmbazel/jobs/lint-viccuad-rm-bazel-ci/builds/5
We should fix that to speed up ci runs, or to not fail on GH blocking us after several downloads. But it doesn't need to block this card.

mook-as commented on .concourse/pipeline.yaml.gomplate:
```diff
@@ -343,8 +343,7 @@ jobs:
         source:
-          repository: thulioassis/bazel-docker-image
-          tag: 2.0.0
+          repository: splatform/catapult
```
Shouldn't this be pinned to _something_? Otherwise when we ship a new version (e.g. for kubecf 2.8) it will break older pipelines (e.g. kubecf 2.4).

viccuad replied:
I agree. Found the pipeline that releases catapult images, started tagging there:
  https://github.com/SUSE/cloudfoundry/pull/180
  https://concourse.suse.de/teams/main/pipelines/catapult

Added a tag and anchored it throughout the pipeline.

mook-as commented on .concourse/pipeline.yaml.gomplate:
```diff
@@ -353,11 +352,7 @@ jobs:
-          ./dev/linters/helmlint.sh
-          bazel test //rules/kubecf:create_sample_values_test
-          bazel test //deploy/helm/kubecf:values_doc_test
+          make lint
```
Does `make -C kubecf-{{ $branch }} lint` work? If yes, we won't need the `/bin/bash` anymore, I think?

viccuad replied:
True, changed. Leaving the `/bin/bash` for build step, making a one-liner there muddies things.

mook-as commented on .concourse/pipeline.yaml.gomplate:
```diff
@@ -418,8 +413,7 @@ jobs:
         source:
-          repository: thulioassis/bazel-docker-image
-          tag: 2.0.0
+          repository: splatform/catapult
```
Same as above: pin to version please.  It might be useful to use YAML anchors here.

viccuad replied:
Done

mook-as commented on .concourse/pipeline.yaml.gomplate:
```diff
@@ -430,7 +424,9 @@ jobs:
           cd kubecf-{{ $branch }}
-          ./dev/build.sh ../output
+          VERSION="$(./scripts/version.sh)"
+          TARGET_FILE=../output/kubecf-${VERSION}.tgz make kubecf-build
```
If we're overriding the target file name anyway, can we not embed the timestamp in the initial file name? Why do we need the `mv` afterwards?

viccuad replied:
True, simplified.

mook-as commented on .concourse/pipeline.yaml.gomplate:
```diff
@@ -1985,8 +1981,7 @@ jobs:
         source:
-          repository: thulioassis/bazel-docker-image
-          tag: 2.0.0
+          repository: splatform/catapult
```
(See above)

viccuad replied:
Done

mook-as replied:
Hmm, the `repository` part is the same as the merged hash right? Do we need the explicit overrides here?

(As far as I can tell all the other uses of the anchor are fine.)

mook-as commented on scripts/kubecf-build.sh:
```diff
@@ -61,6 +61,8 @@ done
 echo "operatorChartUrl: \"${CF_OPERATOR_URL//\{version\}/${CF_OPERATOR_VERSION}}\"" > "${HELM_DIR}/Metadata.yaml"
 
 ruby rules/kubecf/create_sample_values.rb "${HELM_DIR}/values.yaml" "${HELM_DIR}/sample-values.yaml"
+MODE=check ruby rules/kubecf/create_sample_values.rb "${HELM_DIR}/values.yaml" "${HELM_DIR}/sample-values.yaml"
```
Should this be part of lint rather than build?

viccuad replied:
Could be.
This check needs to be run after the `sample-values` creation, which needs a build folder though. It could be added as a target of `make lint`, but then we are duplicating the creation and handling of the temporal folders. 

Maybe this is better solved when we remove Bazel and refactor `rules/`.

jandubois commented:
> The bash build system downloads on its own shellcheck, yamllint, bosh, instead of using any of those binaries from the CI catapult image

It should only do this when you have set `PINNED_TOOLS=1` before you invoke the linters. If you don't set this, then the build scripts will check if the installed version satisfies the minimum version requirements, and if satisfied, will use whatever you have in the `PATH`. I assume `catapult` is setting this, as I don't see any code in your PR doit it.

`PINNED_TOOLS=1` exists to match the Bazel behaviour of always downloading a specific version of the tools (which only worked when all dependencies were specified correctly, otherwise Bazel was also pulling things from `PATH`).

viccuad commented:
Tackled the review commits, flown the pipeline again, https://concourse.suse.dev/teams/main/pipelines/kubecf-rmbazel/jobs/lint-viccuad-rm-bazel-ci/builds/11 onwards.

viccuad commented:
sigh, the make lint call fails, I don't see why: https://concourse.suse.dev/teams/main/pipelines/kubecf-rmbazel/jobs/lint-viccuad-rm-bazel-ci/builds/17
But I need to leave, I will leave this undone, sorry.

mook-as commented on .concourse/pipeline.yaml.gomplate:
```diff
@@ -340,24 +340,18 @@ jobs:
-          ./dev/linters/helmlint.sh
-          bazel test //rules/kubecf:create_sample_values_test
-          bazel test //deploy/helm/kubecf:values_doc_test
+        - kubecf-{{ $branch }}
```
You're missing the `-C` to change directory, I think.

viccuad replied:
Correct, thanks!

viccuad commented:
Thanks for the review!

Fixed make arguments, removed left-over repository key on *catapult-image yaml pointer, added `shellcheck` and `yamllint` to catapult image, and bumped the image's tag being pinned in this pipeline.
